includes/rest-api/Controllers/Version3/class-wc-rest-orders-controller.php

```diff
@@ -34,31 +34,58 @@ class WC_REST_Orders_Controller extends WC_REST_Orders_V2_Controller {
 	 * @return bool
 	 */
 	protected function calculate_coupons( $request, $order ) {
-		if ( ! isset( $request['coupon_lines'] ) || ! is_array( $request['coupon_lines'] ) ) {
+		if ( ! isset( $request['coupon_lines'] ) ) {
 			return false;
 		}
 
-		// Remove all coupons first to ensure calculation is correct.
-		foreach ( $order->get_items( 'coupon' ) as $coupon ) {
-			$order->remove_coupon( $coupon->get_code() );
-		}
+		// Validate input and at the same time generate the coupon objects to apply.
+
+		$coupons   = array();
+		$discounts = new WC_Discounts( $order );
+
+		$current_order_coupons      = array_values( $order->get_coupons() );
+		$current_order_coupon_codes = array_map(
+			function( $coupon ) {
+				return $coupon->get_code();
+			},
+			$current_order_coupons
+		);
 
 		foreach ( $request['coupon_lines'] as $item ) {
-			if ( is_array( $item ) ) {
-				if ( ! empty( $item['id'] ) ) {
-					throw new WC_REST_Exception( 'woocommerce_rest_coupon_item_id_readonly', __( 'Coupon item ID is readonly.', 'woocommerce' ), 400 );
-				}
+			if ( ! empty( $item['id'] ) ) {
+				throw new WC_REST_Exception( 'woocommerce_rest_coupon_item_id_readonly', __( 'Coupon item ID is readonly.', 'woocommerce' ), 400 );
+			}
 
-				if ( empty( $item['code'] ) ) {
-					throw new WC_REST_Exception( 'woocommerce_rest_invalid_coupon', __( 'Coupon code is required.', 'woocommerce' ), 400 );
-				}
+			if ( empty( $item['code'] ) ) {
+				throw new WC_REST_Exception( 'woocommerce_rest_invalid_coupon', __( 'Coupon code is required.', 'woocommerce' ), 400 );
+			}
 
-				$results = $order->apply_coupon( wc_clean( $item['code'] ) );
+			$coupon_code = wc_format_coupon_code( wc_clean( $item['code'] ) );
+			$coupon      = new WC_Coupon( $coupon_code );
 
-				if ( is_wp_error( $results ) ) {
-					throw new WC_REST_Exception( 'woocommerce_rest_' . $results->get_error_code(), $results->get_error_message(), 400 );
+			// Skip check if the coupon is already applied to the order, as this could wrongly throw an error for single-use coupons.
+			if ( ! in_array( $coupon_code, $current_order_coupon_codes, true ) ) {
+				$check_result = $discounts->is_coupon_valid( $coupon );
+				if ( is_wp_error( $check_result ) ) {
+					throw new WC_REST_Exception( 'woocommerce_rest_' . $check_result->get_error_code(), $check_result->get_error_message(), 400 );
 				}
 			}
+
+			$coupons[] = $coupon;
+		}
+
+		// Remove all coupons first to ensure calculation is correct.
+		foreach ( $order->get_items( 'coupon' ) as $existing_coupon ) {
+			$order->remove_coupon( $existing_coupon->get_code() );
+		}
+
+		// Apply the coupons.
+		foreach ( $coupons as $new_coupon ) {
+			$results = $order->apply_coupon( $new_coupon );
```

Comment:
looks like sending `new_coupon` object is causing a cache invalidation problem because props like `usage_count` etc are not refreshed even if they are changed decreased in line 79. 

we should instead send `new_coupon->get_code()` like before to force a refresh from DB.

Reply:
Done!